compiler/src/iree/compiler/PluginAPI/PluginManager.h

```diff
@@ -0,0 +1,233 @@
+struct PluginManagerSession;
+class PluginRegistrar;
+
+// Registration functions are exported with this signature.
+using PluginRegistrationFunction = bool (*)(PluginRegistrar *);
+
+// Empty options class that satisfies the contract for an OptionsBinder.
+// Used by default if a plugin does not support options.
+struct EmptyPluginOptions {
+  static void bindOptions(OptionsBinder &binder) {}
+};
+
+// Abstract class representing a plugin registration. It is responsible for
+// various global initialization and creation of plugin sessions that mirror
+// the lifetime of and |iree_compiler_session_t| for when the plugin is
+// activated.
+//
+// This is typically not instantiated directly but via the PluginSession
+// CRTP helper which manages most details.
+class AbstractPluginRegistration {
+ public:
+  AbstractPluginRegistration(std::string pluginId)
+      : pluginId(std::move(pluginId)) {}
+  virtual ~AbstractPluginRegistration();
+
+  // Gets the plugin id. Valid for the life of the registration.
+  std::string_view getPluginId() { return pluginId; }
+
+  // Performs once-only global initialization. This is called prior to any
+  // sessions being created and affects everything in the process. It is
+  // best used for passes and other such things.
+  // The default implementation does nothing.
+  virtual void globalInitialize() {}
+
+  // Initializes the process-global command line interface. This will be called
+  // if the CLI is enabled, and if so, it indicates that creates sessions
+  // must configure any options from the CLI environment.
+  virtual void initializeCLI() {}
+
+  // If a plugin is activated, performs all needed registrations into the
+  // compiler session's DialectRegistry. By default, does nothing.
+  // Note that this does global registration, regardless of whether the
+  // plugin is loaded. It should be used sparingly for things that cannot
+  // change behavior. It is safer to customize the context on a per-session
+  // basis in a plugin session's activate() method (i.e. if registering
+  // interfaces or behavior changes extensions).
+  virtual void registerDialects(DialectRegistry &registry) {}
+
+  // Creates a concrete session. If the CLI was initialized, then this should
+  // also ensure that any command line options were managed properly into
+  // the session instance.
+  virtual std::unique_ptr<AbstractPluginSession> createSession(
+      MLIRContext *context) = 0;
+
+ private:
+  std::string pluginId;
+};
+
+// Primary base class that plugins extend to provide functionality and
+// extensions to the compiler.
+//
+// A plugin session's life-cycle is bound to an |iree_compiler_session_t| for
+// which it is activated (typically, a CLI will only have a single session, but
+// APIs can create as many as they want within the same process).
+//
+// Most users will inherit from this class via the PluginSession CRTP helper,
+// which adds some niceties and static support for command line option
```

Comment:
static support here threw me off, I was afraid I'd run into a static initializer kind thing.

Reply:
Well LLVM cl is all static initializer stuff. But this carefully makes it work anyway. s/static/global/